### ISSUE TITLE
Don't apply tail-call transformer in sloppy mode

### ISSUE THREAD

**Issue by anba** [google/traceur-compiler#1745]
The tail-call transformer should only be applied for strict mode code.

Compiling `/tmp/i.js` with `./traceur --proper-tail-calls=true --out /tmp/o.js --script /tmp/i.js`:
```javascript
function f() {
  return f();
}
```
Produces this output:
```javascript
var $__1 = $traceurRuntime.initTailRecursiveFunction(f);
function f() {
  return $traceurRuntime.call(function() {
    return $traceurRuntime.continuation(f, null, []);
  }, this, arguments);
}
```

**Comment by arv** [comment]
@mnieper

**Comment by mnieper** [comment]
Yeah, there is currently no check for that in the code. @arv: Does a FunctionBody knows whether it is in strict mode or not?

**Comment by mnieper** [comment]
@arv Could `Parser.js` add a flag to a `FunctionBody` according whether `strictMode_` or not?

**Comment by arv** [comment]
I've been thinking about this before. Having the language mode as part of the AST would definitely be useful. The downside is that it has to be passed along in all transformers which I think is going to be a major hurdle at this point.

We have a ScopeChainBuilder that does keep track of the language mode.

Another option would be to have a base transformer that tracks the language mode. Since the AsyncGeneratorTransformer is a TempVarTransformer  it would have to be a subclass of TempVarTransformer (traits/mixins missing in action).
